// @openzeppelin/platform-deploy-client/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VERSION = void 0;
var api_1 = require("./api");
Object.defineProperty(exports, "PlatformClient", { enumerable: true, get: function () { return api_1.PlatformClient; } });
var block_explorer_api_key_1 = require("./api/block-explorer-api-key");
Object.defineProperty(exports, "BlockExplorerApiKeyClient", { enumerable: true, get: function () { return block_explorer_api_key_1.BlockExplorerApiKeyClient; } });
var deployment_1 = require("./api/deployment");
Object.defineProperty(exports, "DeploymentClient", { enumerable: true, get: function () { return deployment_1.DeploymentClient; } });
var upgrade_1 = require("./api/upgrade");
Object.defineProperty(exports, "UpgradeClient", { enumerable: true, get: function () { return upgrade_1.UpgradeClient; } });
var deployment_config_1 = require("./api/deployment-config");
Object.defineProperty(exports, "DeploymentConfigClient", { enumerable: true, get: function () { return deployment_config_1.DeploymentConfigClient; } });
__exportStar(require("./models"), exports);
// eslint-disable-next-line @typescript-eslint/no-var-requires
exports.VERSION = require('../package.json').version;
